### ISSUE TITLE
Redesign PhET-iO API and implementation

### ISSUE THREAD

**Issue by samreid** [phetsims/faradays-law#79]
In https://github.com/phetsims/phet-io/issues/1178#issuecomment-329573907 we discussed changing our PhET-iO philosophy to be less coupled to implementation and it was suggested we take the ideas for a test drive on a simple sim like Faraday's Law.  Ideally this would start with feedback from designers about what we would like clients to be able to do with the simulation.

@ariel-phet can you get feedback from designers?  This could be in a design meeting but may be just as efficient or more efficient over email/issue comments.  The end result would be a table like the one we created for Build an Atom: https://github.com/phetsims/build-an-atom/issues/156#issuecomment-324743529

**Comment by ariel-phet** [comment]
@samreid can this wait until next week design meeting? Would be good to have both Kathy and Amy available.

**Comment by samreid** [comment]
That would be fine.

**Comment by zepumph** [comment]
A few "leaking implementation details" that @samreid and I found while working on https://github.com/phetsims/phet-io-wrappers/issues/30:

* There are two PhETMenuButton tandems per item. One on the home screen, and one on the navigation bar.
* There are started/endedCallbacksForFiredEmitters on every button, are all of these necessary?

**Comment by ariel-phet** [comment]
| Feature        | Value           | Complexity  | Complete  | 
| ------------- |:-------------:| -----:| :-----:| 
 |  | | |  |
 |  | | |  |
 |  | | | ✓ |

**Comment by samreid** [comment]
We used the "feature / value / complexity / complete" pattern in Build an Atom game, but it turned out that we only really needed the feature/complete columns.  By that token, it would be sufficient to use a checklist.  If developers run into something that is too complex, we can revisit then.

**Comment by zepumph** [comment]
From the design meeting today, we went through Faraday's law twice, once looking at the sim, and a second time looking at instance proxies, to see what items should be part of the stable, supported api. 

Here are the objects we instrument moving forward:

**Control Panel**

Toggle visibility on the following items, and make sure all of the related model properties are instrumented so they can be set through the api
- [ ] Field lines check box
- [ ] The show coils radio button group (not individual buttons)
- [ ] Reset all button (in all sims)
- [ ] Flip magnet button

**Main Sim**

Toggle visibility
- [ ] Voltage meter
- [ ] Group of arrows around the magnet(as a single Node)
- [ ] Individual arrows

Other
- [ ] customize the position of the magnet to anywhere on the screen
- [ ] instrument the drag handler
- [ ] Important model properties like theta/signal.
- [ ] Instrument the voltage text for visibility/custom text.

In terms of removing tandems that are already there, here is a list of most of them. I will want to do another pass to make sure all looks good
- [ ] Voltmeter needle.
- [ ] Started/EndedCallbacksForFiredEmitters on the reset all button
- [ ] singleCoilRadioButton
- [ ] doubleCoilRadioButton

This doesn't include tandems for Joist/Scenery/A11y stuff. Of 54 tandems in FL instance proxies right now, 29 of them are in this category. I think that another design meeting is necessary for design around these features that apply to all sims.

**Comment by samreid** [comment]
The arrow nodes on the bar magnet are instrumented separately.  Our options are to:

a. put them into a similar parent node (would automatically work in instance proxies and have a familiar api)
b. add a TType that has a setArrowsVisible method (would not work automatically in instance proxies but wouldn't require reworking the sim).

I'm leaning toward investigating (a) but wanted to run it past @zepumph before I get too far.  Thoughts?

**Comment by samreid** [comment]
When the voltmeter is hidden, we would also like to hide the voltmeter wires.  It seems we won't be easily able to put them in the same node due to z-layering:

```js
    // wires
    this.addChild( new CoilsWiresNode( this.aligner, model.showSecondCoilProperty ) );
    this.addChild( new VoltmeterWiresNode( this.aligner, voltmeterNode ) );

    // bulb added
    this.addChild( bulbNode );

    // coils added
    this.addChild( bottomCoilNode );
    this.addChild( topCoilNode );
    model.showSecondCoilProperty.linkAttribute( topCoilNode, 'visible' );

    // control panel
    var controlPanelNode = new ControlPanelNode( model, tandem.createTandem( 'controlPanel' ) );
    this.addChild( controlPanelNode );

    // voltmeter added
    voltmeterNode.center = this.aligner.voltmeterPosition;
    this.addChild( voltmeterNode );
```

How about if the wires listen to the visibility of the voltmeter node and update themselves when it is hidden?

**Comment by samreid** [comment]
Synchronizing voltmeter wires and voltmeter visibility implemented as prescribed.

**Comment by samreid** [comment]
@zepumph at the design meeting did anyone mention wanting to hide the reset all button?

**Comment by samreid** [comment]
Do we need to add a light bulb brightness property for the data stream?  More generally, our design meeting was for customizability--do we need another design meeting for data stream?

**Comment by samreid** [comment]
There are 2 ways to hide the arrows:

faradaysLaw.faradaysLawScreen.model.showMagnetArrowsProperty: TProperty
and
faradaysLaw.faradaysLawScreen.view.magnet.arrows: TNode

It seems it would be best if there was only one way to customize this.  It seems the model value belongs in the data stream if we follow the pattern of view element => model element.

**Comment by samreid** [comment]
I guess it makes sense to go with the TProperty in this case, though down the road we may have nodes without corresponding visible properties, etc, so it will become haphazard.  But generally we should operate on models when possible and use MVC.

**Comment by samreid** [comment]
You can now use `?prefixes=faradaysLaw.faradaysLawScreen` to show these prefixes and filter out the joist/a11y ones which have not yet been designed.

**Comment by samreid** [comment]
I'm going to uninstrument the magnet node (it seems the magnet should always be visible for this sim to be useful).

**Comment by samreid** [comment]
After the above changes, here is the screen API (suppressed faradaysLaw.faradaysLawScreen from each variable):

model.magnetModel.flippedProperty: TProperty.<TBoolean>
model.magnetModel.positionProperty: TProperty.<TVector2>
model.magnetModel.showFieldLinesProperty: TProperty.<TBoolean>
model.showMagnetArrowsProperty: TProperty.<TBoolean>
model.showSecondCoilProperty: TProperty.<TBoolean>
model.voltmeterModel.signalProperty: TProperty.<TNumber>
model.voltmeterModel.thetaProperty: TProperty.<TNumber>
view.controlPanel.flipMagnetButton: TPushButton
view.controlPanel.radioButtonGroup: TNode
view.controlPanel.resetAllButton: TResetAllButton
view.controlPanel.showFieldCheckBox: TCheckBox
view.magnet.dragHandler: TSimpleDragHandler
view.voltmeterNode: TNode
view.voltmeterNode.label: TText

**Comment by samreid** [comment]
Unfortunately, I had to add in the singleCoilRadioButton and doubleCoilRadioButton so they would make sense in the data stream.  But we should find a way to suppress them from instance proxies/customizability.

**Comment by samreid** [comment]
The data stream is much nicer now that I've removed the `thetaProperty` which represented the continuous angle of the needle.

**Comment by samreid** [comment]
It would be good to address the joist issues before republishing this as a stable version.

**Comment by samreid** [comment]
We also need to accomplish https://github.com/phetsims/phet-io/issues/1234 before we can republish this sim as a stable phet-io version.

**Comment by zepumph** [comment]
- [ ] https://github.com/phetsims/faradays-law/issues/83 must be done before republish

**Comment by samreid** [comment]
I'm also working on the code review in #86

**Comment by samreid** [comment]
Back on hold until https://github.com/phetsims/phet-io/issues/1234 and https://github.com/phetsims/faradays-law/issues/83 are complete.

**Comment by zepumph** [comment]
@samreid I don't think that we should consider #83 as part of this. I think of this stable api update having little to do with a11y. We made efforts to get a11y out of phet-io when a11y is not enabled, and faradays-law has had little a11y work done on it. Especially since https://github.com/phetsims/phet-io/issues/1205 is no where near complete, I think that adding a11y to the stable api will come later, when we are ready to advertise a sim as phet-io+accessible. What do you think?

**Comment by zepumph** [comment]
I noted this in https://github.com/phetsims/joist/issues/445#issuecomment-346216200 (the most recent joist+a11y checklist), but not over here, so thanks for keeping me honest!

**Comment by samreid** [comment]
OK sounds good, thanks!

**Comment by samreid** [comment]
Here is the current API:

```json
{
  "instances": {
    "phetio": {
      "phetioTypeName": "PhETIO",
      "phetioState": false,
      "phetioReadOnly": false,
      "phetioInstanceDocumentation": "Central point for PhET-iO interoperability."
    },
    "faradaysLaw.phetioCommandProcessor": {
      "phetioTypeName": "PhetioCommandProcessorIO",
      "phetioState": false,
      "phetioReadOnly": false,
      "phetioInstanceDocumentation": "Processes messages from another frame and returns the results"
    },
    "faradaysLaw.faradaysLawScreen.activeProperty": {
      "phetioTypeName": "PropertyIO.<BooleanIO>",
      "phetioState": true,
      "phetioReadOnly": true,
      "phetioInstanceDocumentation": "Indicates whether the screen is active.  For single-screen simulations, the screen is always active."
    },
    "faradaysLaw": {
      "phetioTypeName": "SimIO",
      "phetioState": false,
      "phetioReadOnly": false,
      "phetioInstanceDocumentation": "Represents the entire simulation"
    },
    "faradaysLaw.screenIndexProperty": {
      "phetioTypeName": "NumberPropertyIO",
      "phetioState": true,
      "phetioReadOnly": false,
      "phetioInstanceDocumentation": "Indicates which screen is selected (0-indexed)"
    },
    "faradaysLaw.activeProperty": {
      "phetioTypeName": "PropertyIO.<BooleanIO>",
      "phetioState": true,
      "phetioReadOnly": false,
      "phetioInstanceDocumentation": "Indicates whether the entire simulation is running and processing user input"
    },
    "faradaysLaw.browserTabVisibleProperty": {
      "phetioTypeName": "PropertyIO.<BooleanIO>",
      "phetioState": true,
      "phetioReadOnly": true,
      "phetioInstanceDocumentation": "Indicates whether the browser tab containing the simulation is currently visible"
    },
    "faradaysLaw.phetButtonPickableProperty": {
      "phetioTypeName": "PropertyIO.<BooleanIO>",
      "phetioState": true,
      "phetioReadOnly": false,
      "phetioInstanceDocumentation": "Determines whether the PhET Button in the bottom right of the screen can be pressed, and applies to the PhET Button on the home screen (if any) and navigation bar."
    },
    "faradaysLaw.navigationBar.titleTextNode": {
      "phetioTypeName": "TextIO",
      "phetioState": true,
      "phetioReadOnly": false,
      "phetioInstanceDocumentation": "Displays the title of the simulation in the navigation bar (bottom left)"
    },
    "faradaysLaw.navigationBar.phetButton.phetMenu.screenshotMenuItem": {
      "phetioTypeName": "MenuItemIO",
      "phetioState": true,
      "phetioReadOnly": false,
      "phetioInstanceDocumentation": "This menu item captures a screenshot from the simulation and saves it to the file system"
    },
    "faradaysLaw.navigationBar.phetButton.phetMenu.fullScreenMenuItem": {
      "phetioTypeName": "MenuItemIO",
      "phetioState": true,
      "phetioReadOnly": false,
      "phetioInstanceDocumentation": "This menu item requests full-screen access for the simulation display"
    },
    "faradaysLaw.navigationBar.phetButton.phetMenu.aboutMenuItem": {
      "phetioTypeName": "MenuItemIO",
      "phetioState": true,
      "phetioReadOnly": true,
      "phetioInstanceDocumentation": "This menu item shows a dialog with information about the simulation"
    },
    "faradaysLaw.navigationBar.phetButton.phetMenu": {
      "phetioTypeName": "PhetMenuIO",
      "phetioState": false,
      "phetioReadOnly": false,
      "phetioInstanceDocumentation": "This menu is displayed when the PhET button is pressed"
    },
    "faradaysLaw.navigationBar.phetButton": {
      "phetioTypeName": "JoistButtonIO",
      "phetioState": false,
      "phetioReadOnly": false,
      "phetioInstanceDocumentation": "The button that appears in the bottom of the right of the screen, which shows a menu when pressed."
    },
    "faradaysLaw.faradaysLawScreen.model.showTopCoilProperty": {
      "phetioTypeName": "PropertyIO.<BooleanIO>",
      "phetioState": true,
      "phetioReadOnly": false,
      "phetioInstanceDocumentation": ""
    },
    "faradaysLaw.faradaysLawScreen.model.showMagnetArrowsProperty": {
      "phetioTypeName": "PropertyIO.<BooleanIO>",
      "phetioState": true,
      "phetioReadOnly": false,
      "phetioInstanceDocumentation": ""
    },
    "faradaysLaw.faradaysLawScreen.model.magnet.positionProperty": {
      "phetioTypeName": "PropertyIO.<Vector2IO>",
      "phetioState": true,
      "phetioReadOnly": false,
      "phetioInstanceDocumentation": ""
    },
    "faradaysLaw.faradaysLawScreen.model.magnet.orientationProperty": {
      "phetioTypeName": "PropertyIO.<StringIO>",
      "phetioState": true,
      "phetioReadOnly": false,
      "phetioInstanceDocumentation": ""
    },
    "faradaysLaw.faradaysLawScreen.model.magnet.showFieldLinesProperty": {
      "phetioTypeName": "PropertyIO.<BooleanIO>",
      "phetioState": true,
      "phetioReadOnly": false,
      "phetioInstanceDocumentation": ""
    },
    "faradaysLaw.faradaysLawScreen.model.voltmeter.voltageProperty": {
      "phetioTypeName": "NumberPropertyIO",
      "phetioState": true,
      "phetioReadOnly": true,
      "phetioInstanceDocumentation": ""
    },
    "faradaysLaw.faradaysLawScreen.model.voltmeter.signalProperty": {
      "phetioTypeName": "NumberPropertyIO",
      "phetioState": true,
      "phetioReadOnly": false,
      "phetioInstanceDocumentation": ""
    },
    "faradaysLaw.faradaysLawScreen.view.voltmeterNode": {
      "phetioTypeName": "NodeIO",
      "phetioState": true,
      "phetioReadOnly": false,
      "phetioInstanceDocumentation": ""
    },
    "faradaysLaw.faradaysLawScreen.view.voltmeterNode.label": {
      "phetioTypeName": "TextIO",
      "phetioState": true,
      "phetioReadOnly": false,
      "phetioInstanceDocumentation": ""
    },
    "faradaysLaw.faradaysLawScreen.view.resetAllButton": {
      "phetioTypeName": "PushButtonIO",
      "phetioState": true,
      "phetioReadOnly": false,
      "phetioInstanceDocumentation": ""
    },
    "faradaysLaw.faradaysLawScreen.view.resetAllButton.isFiringProperty": {
      "phetioTypeName": "DerivedPropertyIO.<BooleanIO>",
      "phetioReadOnly": false,
      "phetioInstanceDocumentation": ""
    },
    "faradaysLaw.faradaysLawScreen.view.flipMagnetButton": {
      "phetioTypeName": "PushButtonIO",
      "phetioState": true,
      "phetioReadOnly": false,
      "phetioInstanceDocumentation": ""
    },
    "faradaysLaw.faradaysLawScreen.view.coilRadioButtonGroup.singleCoilRadioButton": {
      "phetioTypeName": "RadioButtonGroupMemberIO",
      "phetioState": true,
      "phetioReadOnly": false,
      "phetioInstanceDocumentation": ""
    },
    "faradaysLaw.faradaysLawScreen.view.coilRadioButtonGroup.doubleCoilRadioButton": {
      "phetioTypeName": "RadioButtonGroupMemberIO",
      "phetioState": true,
      "phetioReadOnly": false,
      "phetioInstanceDocumentation": ""
    },
    "faradaysLaw.faradaysLawScreen.view.coilRadioButtonGroup": {
      "phetioTypeName": "NodeIO",
      "phetioState": true,
      "phetioReadOnly": false,
      "phetioInstanceDocumentation": ""
    },
    "faradaysLaw.faradaysLawScreen.view.showFieldCheckbox": {
      "phetioTypeName": "CheckboxIO",
      "phetioState": true,
      "phetioReadOnly": false,
      "phetioInstanceDocumentation": ""
    },
    "faradaysLaw.faradaysLawScreen.view.magnet.draggableNode": {
      "phetioTypeName": "NodeIO",
      "phetioState": true,
      "phetioReadOnly": false,
      "phetioInstanceDocumentation": ""
    },
    "faradaysLaw.faradaysLawScreen.view.magnet.dragHandler.isDraggedProperty": {
      "phetioTypeName": "PropertyIO.<BooleanIO>",
      "phetioState": true,
      "phetioReadOnly": false,
      "phetioInstanceDocumentation": ""
    },
    "faradaysLaw.faradaysLawScreen.view.magnet.dragHandler": {
      "phetioTypeName": "SimpleDragHandlerIO",
      "phetioState": false,
      "phetioReadOnly": false,
      "phetioInstanceDocumentation": ""
    },
    "faradaysLaw.barrierRectangle": {
      "phetioTypeName": "BarrierRectangleIO",
      "phetioState": false,
      "phetioReadOnly": true,
      "phetioInstanceDocumentation": ""
    }
  }
}
```

**Comment by samreid** [comment]
We'll need to address https://github.com/phetsims/sun/issues/336 before we can make an RC.

**Comment by samreid** [comment]
After the above commit, every instance has `phetioInstanceDocumentation`

**Comment by ariel-phet** [comment]
Not aware of anymore comments I need to make here (looked over API).

**Comment by zepumph** [comment]
The current working list for FL stable release is https://github.com/issues?utf8=%E2%9C%93&q=is%3Aopen+is%3Aissue+label%3Ameeting%3Aphet-io+user%3Aphetsims+label%3Apriority%3A2-high+

Currently there are 10 issues in that list.

**Comment by zepumph** [comment]
This issue is largely out of date, and we will likely need to review the whole sim (even briefly) based on our newest phet-io implementation and design patterns. I'm going to close this. We can pick things up in https://github.com/phetsims/faradays-law/issues/158
